mantidproject/mantid
# PolDiffILL: single crystal reduction

This PR adds support to reduce single-crystal data measured by D7 instrument at the ILL in `PolDiffILLReduction` and `D7AbsoluteCrossSections` algorithms.

- [x] Optimise the processing time of 400+ files of an omega-scan. Achieved with replacing `Load` with `LoadAndMerge` and conjoining workspaces provided by `PolDiffILLReduction` into 2D workspaces containing the entire omega scan. 
- [x] Add anisotropic magnetism cross-section separation based on _W. Schweika 2010 J. Phys.: Conf. Ser. 211 012026_.
- [x] Output the single crystal reduced data as a mapped on a Qx - Qy plane, reimplementingLAMP's QRebin method.
- [x] Verify that the separated cross-sections are comparable with results from LAMP.

Recalculation of X and Y axes values can be done outside of the reduction and normalisation algorithms, for example as a part of the reduction script.

**To test:**

1. Check the code quality
2. Compile and read new documentation, in particular formulae and explanations regarding the anisotropic magnetism treatment
3. Run an example reduction script, see below:

If you have access to ILL data, the full reduction is based on exp_5-32-870 from cycle 193. The reduction below is a simplified reduction based only on a small part of the full dataset. Make sure you have SystemTest data with ILL/D7 in your datasearch directories.  

```
vanadium_mass = 8.535
sample_formula_mass = 137.33 * 2.0 + 54.93 + 127.6 + 15.999 * 6.0
sample_mass = 7.83
vanadium_dictionary = {'SampleMass': vanadium_mass, 'FormulaUnits': 1, 'FormulaUnitMass': 50.942}
sample_dictionary = {'SampleMass': sample_mass, 'FormulaUnits': 1, 'FormulaUnitMass': sample_formula_mass,
                                  'KiXAngle': 45.0, 'OmegaShift': 52.5}
calibration_file = "D7_YIG_calibration.xml"

# Empty container for quartz and vanadium
PolDiffILLReduction(
            Run='450747:450748',
            OutputWorkspace='container_ws',
            ProcessAs='Empty'
)

# Empty container for bank position 1 (bt1), tth=79.5
PolDiffILLReduction(
            Run='397406:397407',
            OutputTreatment='AveragePol',
            OutputWorkspace='container_bt1_ws',
            ProcessAs='Empty'
)
# empty container for bt2, tth=75
        PolDiffILLReduction(
            Run='397397:397398',
            OutputTreatment='AveragePol',
            OutputWorkspace='container_bt2_ws',
            ProcessAs='Empty'
        )

        PolDiffILLReduction(
            Run='450769:450770',
            OutputWorkspace='pol_corrections',
            EmptyContainerWorkspace='container_ws',
            Transmission='0.9',
            OutputTreatment='AveragePol',
            ProcessAs='Quartz'
        )

        PolDiffILLReduction(
            Run='450835:450836',
            OutputWorkspace='vanadium_ws',
            EmptyContainerWorkspace='container_ws',
            Transmission='0.89',
            QuartzWorkspace='pol_corrections',
            OutputTreatment='Sum',
            SampleGeometry='None',
            SelfAttenuationMethod='Transmission',
            SampleAndEnvironmentProperties=vanadium_dictionary,
            AbsoluteNormalisation=True,
            InstrumentCalibration=calibration_file,
            ProcessAs='Vanadium'
        )

        # bank position 1, tth=79.5
        PolDiffILLReduction(
            Run='399451:399452',
            OutputWorkspace='bt1',
            EmptyContainerWorkspace='container_bt1_ws',
            Transmission='0.95',
            QuartzWorkspace='pol_corrections',
            OutputTreatment='Individual',
            SampleGeometry='None',
            SampleAndEnvironmentProperties=sample_dictionary,
            MeasurementTechnique='SingleCrystal',
            InstrumentCalibration=calibration_file,
            ProcessAs='Sample'
        )
        # bank position 2, tth=75
        PolDiffILLReduction(
            Run='400287:400288',
            OutputWorkspace='bt2',
            EmptyContainerWorkspace='container_bt2_ws',
            Transmission='0.95',
            QuartzWorkspace='pol_corrections',
            OutputTreatment='Individual',
            SampleGeometry='None',
            SampleAndEnvironmentProperties=sample_dictionary,
            MeasurementTechnique='SingleCrystal',
            InstrumentCalibration=calibration_file,
            ProcessAs='Sample'
        )
        appended_ws = 'appended_ws'
        AppendSpectra(InputWorkspace1='bt1', InputWorkspace2='bt2',
                      OutputWorkspace=appended_ws)
        # names need to be re-set, AppendSpectra just concatenates them
        possible_polarisations = ['ZPO_ON', 'ZPO_OFF', 'XPO_ON', 'XPO_OFF', 'YPO_ON', 'YPO_OFF']
        polarisation = ""
        for entry in mtd[appended_ws]:
            entry_name = entry.name()
            for polarisation in possible_polarisations:
                if polarisation in entry_name:
                    break
            RenameWorkspace(InputWorkspace=entry, OutputWorkspace="{}_{}".format(appended_ws, polarisation))

        D7AbsoluteCrossSections(
            InputWorkspace=appended_ws,
            OutputWorkspace='normalized_single_crystal_XYZ',
            CrossSectionSeparationMethod='XYZ',
            NormalisationMethod='Vanadium',
            VanadiumInputWorkspace='vanadium_ws',
            OutputUnits='Qxy',
            SampleAndEnvironmentProperties=sample_dictionary,
            AbsoluteUnitsNormalisation=False,
            IsotropicMagnetism=False,
            MeasurementTechnique='SingleCrystal',
            ClearCache=True
        )
```

Fixes #31419 . 

Release and documentation updated.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
